framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -2183,6 +2183,25 @@ public AnnotatedDeclaredType getEnclosingType(TypeElement typeElement, Tree tree
     return thisType;
   }
 
+  /**
+   * Returns the super {@link AnnotatedTypeMirror} of the enclosing type at the location of {@code
+   * tree} that is the same type as {@code typeElement}.
+   *
+   * @param typeElement type of the enclosing type to return
+   * @param tree location to use
+   * @return the super enclosing type at the location of {@code tree} that is the same type as
+   *     {@code typeElement}
+   */
+  public AnnotatedDeclaredType getEnclosingSuperType(TypeElement typeElement, Tree tree) {
```

Comment:
I find the contract of `getEnclosingSuperType` confusing.  I suggest changing it to just return the matching enclosing type (that's simpler and feels more like a logical unit), and having the client compute the supertype.